{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Applying logistic regression in python to discriminate between control and M. bovis infected animals using peripheral blood transcriptomics data from abdelaal et al., 2020\n",
    "### This analysis considers two approaches, one using logistic regression on variable genes that have been preprocessed using DESeq2 (vst normalised) and the other using latent variables inferred using PCA, ICA and NMF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load in all necessary packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.stats import kurtosis\n",
    "\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import classification_report, confusion_matrix, ConfusionMatrixDisplay,roc_curve,auc, make_scorer,mean_squared_error\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import KFold, GridSearchCV, cross_val_score, PredefinedSplit\n",
    "from sklearn.decomposition import PCA, NMF, FastICA\n",
    "from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score\n",
    "from sklearn.pipeline import Pipeline\n",
    "import warnings\n",
    "import pickle\n",
    "from sklearn.pipeline import Pipeline\n",
    "import matplotlib\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "#font for plots\n",
    "font = {'fontname':'Arial'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "abdelaal_data_raw = pd.read_csv(\"/home/workspace/jogrady/ML4TB/work/normalisation/vst_individual/abdelaal_vst_normalised_data.txt\", sep = \"\\t\").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "abdelaal_data_raw = pd.read_csv(\"/home/workspace/jogrady/ML4TB/work/normalisation/vst_individual/abdelaal_vst_normalised_data.txt\", sep = \"\\t\").T\n",
    "abdelaal_labels = pd.read_csv(\"/home/workspace/jogrady/ML4TB/data/abdelaal/abdelaal_samples.csv\", sep = \"\\t\")\n",
    "abdelaal_labels = abdelaal_labels[[\"Animal_Code\", \"Week\", \"Status\"]].drop_duplicates()\n",
    "\n",
    "\n",
    "\n",
    "np.array(abdelaal_labels[\"Animal_Code\"]) == np.array(abdelaal_data_raw.index)\n",
    "abdelaal_labels[\"ID\"] = abdelaal_labels[\"Animal_Code\"].str.extract(r\"^(.*)_[^_]+$\", expand=False)\n",
    "abdelaal_labels[\"Fold\"] = abdelaal_labels.groupby(\"ID\").ngroup()\n",
    "abdelaal_labels.loc[abdelaal_labels[\"Week\"] == \"W0\", \"Status\"] = \"Control\" # Note the infected samples were sampled after infection (immediately) so may be worthwhile labelling them as Control (note this increased the CV accuracy from ~ 50 to 80)\n",
    "# Check right order\n",
    "np.array(abdelaal_labels[\"Animal_Code\"]) == np.array(abdelaal_data_raw.index)\n",
    "\n",
    "abdelaal_labels_simple = abdelaal_labels[\"Status\"].to_numpy()\n",
    "abdelaal_folds = abdelaal_labels[\"Fold\"].to_numpy()\n",
    "\n",
    "\n",
    "abdelaal_labels_simple = np.where(abdelaal_labels_simple == \"Control\", 0, np.where(abdelaal_labels_simple == \"Infected\", 1, abdelaal_labels_simple))\n",
    "abdelaal_labels_simple = abdelaal_labels_simple.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Animal_Code</th>\n",
       "      <th>Week</th>\n",
       "      <th>Status</th>\n",
       "      <th>ID</th>\n",
       "      <th>Fold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Infected_1_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Infected_1_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Infected_2_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Infected_2_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Infected_3_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Infected_3_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Infected_4_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Infected_4_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Infected_5_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Infected_5_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Infected_6_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Infected_6_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Infected</td>\n",
       "      <td>Infected_6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Uninfected_1_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Uninfected_1_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Uninfected_2_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Uninfected_2_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Uninfected_3_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Uninfected_3_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Uninfected_4_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Uninfected_4_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Uninfected_5_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Uninfected_5_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Uninfected_6_20</td>\n",
       "      <td>W20</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_6</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Uninfected_6_8</td>\n",
       "      <td>W8</td>\n",
       "      <td>Control</td>\n",
       "      <td>Uninfected_6</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Animal_Code Week    Status            ID  Fold\n",
       "0     Infected_1_20  W20  Infected    Infected_1     0\n",
       "2      Infected_1_8   W8  Infected    Infected_1     0\n",
       "4     Infected_2_20  W20  Infected    Infected_2     1\n",
       "6      Infected_2_8   W8  Infected    Infected_2     1\n",
       "8     Infected_3_20  W20  Infected    Infected_3     2\n",
       "10     Infected_3_8   W8  Infected    Infected_3     2\n",
       "12    Infected_4_20  W20  Infected    Infected_4     3\n",
       "14     Infected_4_8   W8  Infected    Infected_4     3\n",
       "16    Infected_5_20  W20  Infected    Infected_5     4\n",
       "18     Infected_5_8   W8  Infected    Infected_5     4\n",
       "20    Infected_6_20  W20  Infected    Infected_6     5\n",
       "22     Infected_6_8   W8  Infected    Infected_6     5\n",
       "24  Uninfected_1_20  W20   Control  Uninfected_1     6\n",
       "26   Uninfected_1_8   W8   Control  Uninfected_1     6\n",
       "28  Uninfected_2_20  W20   Control  Uninfected_2     7\n",
       "30   Uninfected_2_8   W8   Control  Uninfected_2     7\n",
       "32  Uninfected_3_20  W20   Control  Uninfected_3     8\n",
       "34   Uninfected_3_8   W8   Control  Uninfected_3     8\n",
       "36  Uninfected_4_20  W20   Control  Uninfected_4     9\n",
       "38   Uninfected_4_8   W8   Control  Uninfected_4     9\n",
       "40  Uninfected_5_20  W20   Control  Uninfected_5    10\n",
       "42   Uninfected_5_8   W8   Control  Uninfected_5    10\n",
       "44  Uninfected_6_20  W20   Control  Uninfected_6    11\n",
       "46   Uninfected_6_8   W8   Control  Uninfected_6    11"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abdelaal_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calcualte variances for VST normalised genes\n",
    "variances = abdelaal_data_raw.var(axis=0)\n",
    "# take top 20% and filter\n",
    "threshold = variances.quantile(.80) \n",
    "genes = variances > threshold\n",
    "genes= genes.loc[genes==True].index\n",
    "abdelaal_data = abdelaal_data_raw.filter(items = genes, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSBTAG00000054829</th>\n",
       "      <th>RCAN1</th>\n",
       "      <th>SMIM11</th>\n",
       "      <th>ITSN1</th>\n",
       "      <th>ABCG1</th>\n",
       "      <th>TFF2</th>\n",
       "      <th>TMPRSS3</th>\n",
       "      <th>RSPH1</th>\n",
       "      <th>IL1RAP</th>\n",
       "      <th>P3H2</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSBTAG00000053997</th>\n",
       "      <th>ENSBTAG00000054086</th>\n",
       "      <th>ENSBTAG00000050585</th>\n",
       "      <th>ENSBTAG00000052194</th>\n",
       "      <th>SERPINB4</th>\n",
       "      <th>ENSBTAG00000050153</th>\n",
       "      <th>ENSBTAG00000050608</th>\n",
       "      <th>MAD2L1</th>\n",
       "      <th>ENSBTAG00000054081</th>\n",
       "      <th>ENSBTAG00000049569</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Infected_1_20</th>\n",
       "      <td>5.020736</td>\n",
       "      <td>9.472897</td>\n",
       "      <td>8.265235</td>\n",
       "      <td>9.607765</td>\n",
       "      <td>10.199095</td>\n",
       "      <td>6.423200</td>\n",
       "      <td>5.446811</td>\n",
       "      <td>5.020736</td>\n",
       "      <td>8.269280</td>\n",
       "      <td>6.308950</td>\n",
       "      <td>...</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>8.865097</td>\n",
       "      <td>4.933724</td>\n",
       "      <td>6.391624</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>6.238502</td>\n",
       "      <td>7.581310</td>\n",
       "      <td>6.291721</td>\n",
       "      <td>5.020736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Infected_1_8</th>\n",
       "      <td>4.824982</td>\n",
       "      <td>9.568001</td>\n",
       "      <td>8.278523</td>\n",
       "      <td>9.761322</td>\n",
       "      <td>10.372165</td>\n",
       "      <td>6.454131</td>\n",
       "      <td>5.738425</td>\n",
       "      <td>5.421722</td>\n",
       "      <td>8.782285</td>\n",
       "      <td>5.952382</td>\n",
       "      <td>...</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>10.476969</td>\n",
       "      <td>4.941008</td>\n",
       "      <td>6.338420</td>\n",
       "      <td>4.941008</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>6.500611</td>\n",
       "      <td>7.688112</td>\n",
       "      <td>7.109312</td>\n",
       "      <td>4.824982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Infected_2_20</th>\n",
       "      <td>4.543355</td>\n",
       "      <td>9.083210</td>\n",
       "      <td>8.664853</td>\n",
       "      <td>8.918923</td>\n",
       "      <td>10.963202</td>\n",
       "      <td>5.516665</td>\n",
       "      <td>5.000588</td>\n",
       "      <td>5.070595</td>\n",
       "      <td>8.262210</td>\n",
       "      <td>6.481582</td>\n",
       "      <td>...</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>7.116487</td>\n",
       "      <td>6.936616</td>\n",
       "      <td>5.370276</td>\n",
       "      <td>5.000588</td>\n",
       "      <td>6.467869</td>\n",
       "      <td>5.482480</td>\n",
       "      <td>6.983877</td>\n",
       "      <td>6.337144</td>\n",
       "      <td>4.808073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Infected_2_8</th>\n",
       "      <td>5.016344</td>\n",
       "      <td>9.322084</td>\n",
       "      <td>8.500184</td>\n",
       "      <td>9.392623</td>\n",
       "      <td>10.808786</td>\n",
       "      <td>6.327843</td>\n",
       "      <td>5.209333</td>\n",
       "      <td>5.677687</td>\n",
       "      <td>9.055278</td>\n",
       "      <td>6.423233</td>\n",
       "      <td>...</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>7.195843</td>\n",
       "      <td>6.634089</td>\n",
       "      <td>5.842797</td>\n",
       "      <td>5.355504</td>\n",
       "      <td>5.985045</td>\n",
       "      <td>5.707205</td>\n",
       "      <td>7.254159</td>\n",
       "      <td>6.110927</td>\n",
       "      <td>5.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Infected_3_20</th>\n",
       "      <td>4.924988</td>\n",
       "      <td>9.444165</td>\n",
       "      <td>9.058130</td>\n",
       "      <td>8.918001</td>\n",
       "      <td>11.077527</td>\n",
       "      <td>5.945029</td>\n",
       "      <td>5.501468</td>\n",
       "      <td>5.465131</td>\n",
       "      <td>8.492798</td>\n",
       "      <td>6.445798</td>\n",
       "      <td>...</td>\n",
       "      <td>4.543355</td>\n",
       "      <td>7.618384</td>\n",
       "      <td>7.150779</td>\n",
       "      <td>6.221806</td>\n",
       "      <td>5.081516</td>\n",
       "      <td>6.557695</td>\n",
       "      <td>4.924988</td>\n",
       "      <td>7.116096</td>\n",
       "      <td>5.720759</td>\n",
       "      <td>5.465131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5522 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ENSBTAG00000054829     RCAN1    SMIM11     ITSN1      ABCG1  \\\n",
       "Infected_1_20            5.020736  9.472897  8.265235  9.607765  10.199095   \n",
       "Infected_1_8             4.824982  9.568001  8.278523  9.761322  10.372165   \n",
       "Infected_2_20            4.543355  9.083210  8.664853  8.918923  10.963202   \n",
       "Infected_2_8             5.016344  9.322084  8.500184  9.392623  10.808786   \n",
       "Infected_3_20            4.924988  9.444165  9.058130  8.918001  11.077527   \n",
       "\n",
       "                   TFF2   TMPRSS3     RSPH1    IL1RAP      P3H2  ...  \\\n",
       "Infected_1_20  6.423200  5.446811  5.020736  8.269280  6.308950  ...   \n",
       "Infected_1_8   6.454131  5.738425  5.421722  8.782285  5.952382  ...   \n",
       "Infected_2_20  5.516665  5.000588  5.070595  8.262210  6.481582  ...   \n",
       "Infected_2_8   6.327843  5.209333  5.677687  9.055278  6.423233  ...   \n",
       "Infected_3_20  5.945029  5.501468  5.465131  8.492798  6.445798  ...   \n",
       "\n",
       "               ENSBTAG00000053997  ENSBTAG00000054086  ENSBTAG00000050585  \\\n",
       "Infected_1_20            4.543355            8.865097            4.933724   \n",
       "Infected_1_8             4.543355           10.476969            4.941008   \n",
       "Infected_2_20            4.543355            7.116487            6.936616   \n",
       "Infected_2_8             4.543355            7.195843            6.634089   \n",
       "Infected_3_20            4.543355            7.618384            7.150779   \n",
       "\n",
       "               ENSBTAG00000052194  SERPINB4  ENSBTAG00000050153  \\\n",
       "Infected_1_20            6.391624  4.543355            4.543355   \n",
       "Infected_1_8             6.338420  4.941008            4.543355   \n",
       "Infected_2_20            5.370276  5.000588            6.467869   \n",
       "Infected_2_8             5.842797  5.355504            5.985045   \n",
       "Infected_3_20            6.221806  5.081516            6.557695   \n",
       "\n",
       "               ENSBTAG00000050608    MAD2L1  ENSBTAG00000054081  \\\n",
       "Infected_1_20            6.238502  7.581310            6.291721   \n",
       "Infected_1_8             6.500611  7.688112            7.109312   \n",
       "Infected_2_20            5.482480  6.983877            6.337144   \n",
       "Infected_2_8             5.707205  7.254159            6.110927   \n",
       "Infected_3_20            4.924988  7.116096            5.720759   \n",
       "\n",
       "               ENSBTAG00000049569  \n",
       "Infected_1_20            5.020736  \n",
       "Infected_1_8             4.824982  \n",
       "Infected_2_20            4.808073  \n",
       "Infected_2_8             5.817100  \n",
       "Infected_3_20            5.465131  \n",
       "\n",
       "[5 rows x 5522 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abdelaal_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the custom folds array to a PredefinedSplit object\n",
    "ps = PredefinedSplit(test_fold=abdelaal_folds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make a pipeline for logistic regression and set the paramaters\n",
    "log_pipe = Pipeline(steps=[\n",
    "('scaler', StandardScaler()), # see comment above (in markdown)\n",
    "('classifier', LogisticRegression(max_iter=10000, solver='saga', tol=0.0001, random_state=42))]) # classifier\n",
    "\n",
    "precision_scorer = make_scorer(precision_score, zero_division=1)  # had to modify zero_division as it was giving problems\n",
    "f1_scorer = make_scorer(f1_score)\n",
    "accuracy_scorer = make_scorer(accuracy_score)\n",
    "recall_scorer = make_scorer(recall_score)\n",
    "\n",
    "# Define scoring dictionary for GridSearchCV\n",
    "scoring = {\n",
    "    'accuracy': accuracy_scorer,\n",
    "    #'f1': f1_scorer,\n",
    "    #'precision': precision_scorer,\n",
    "    #'recall': recall_scorer\n",
    "}\n",
    "\n",
    "# Create a parameter grid - we will search through all these combinations\n",
    "param_grid = {\n",
    "    'classifier__penalty': [\"elasticnet\"],\n",
    "    'classifier__l1_ratio': [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1]\n",
    "}\n",
    " \n",
    "# Create GridSearchCV object\n",
    "grid_search = GridSearchCV(log_pipe, param_grid, cv=ps, verbose=3, n_jobs=60, scoring=scoring, refit=\"accuracy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 12 folds for each of 10 candidates, totalling 120 fits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CV 2/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=   9.4s\n",
      "[CV 5/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=   9.2s\n",
      "[CV 10/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=   8.7s\n",
      "[CV 2/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=   9.9s\n",
      "[CV 11/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=   9.6s\n",
      "[CV 11/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  10.9s\n",
      "[CV 3/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  11.1s\n",
      "[CV 10/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  10.6s\n",
      "[CV 7/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  11.2s\n",
      "[CV 6/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  11.9s\n",
      "[CV 1/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  12.1s\n",
      "[CV 4/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  12.5s\n",
      "[CV 8/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  12.5s\n",
      "[CV 6/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  12.8s\n",
      "[CV 6/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  13.1s\n",
      "[CV 8/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  12.4s\n",
      "[CV 12/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  13.3s\n",
      "[CV 6/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  13.5s\n",
      "[CV 9/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.1s\n",
      "[CV 7/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.3s\n",
      "[CV 5/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  13.7s\n",
      "[CV 4/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.8s\n",
      "[CV 1/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.7s\n",
      "[CV 5/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.2s\n",
      "[CV 4/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.5s\n",
      "[CV 12/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.3s\n",
      "[CV 1/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.3s\n",
      "[CV 1/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.8s\n",
      "[CV 9/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.1s\n",
      "[CV 10/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.4s\n",
      "[CV 5/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  15.1s\n",
      "[CV 10/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.2s\n",
      "[CV 11/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  16.0s\n",
      "[CV 2/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  15.3s\n",
      "[CV 9/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.2s\n",
      "[CV 8/12] END classifier__l1_ratio=0.1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.9s\n",
      "[CV 3/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.3s\n",
      "[CV 2/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.3s\n",
      "[CV 1/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  15.5s\n",
      "[CV 3/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.6s\n",
      "[CV 9/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.8s\n",
      "[CV 6/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.1s\n",
      "[CV 4/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  16.2s\n",
      "[CV 7/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  17.0s\n",
      "[CV 3/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  16.2s\n",
      "[CV 12/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  16.4s\n",
      "[CV 8/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.7s\n",
      "[CV 12/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  16.8s\n",
      "[CV 11/12] END classifier__l1_ratio=0.4, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  17.4s\n",
      "[CV 7/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  17.5s\n",
      "[CV 5/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  17.3s\n",
      "[CV 7/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  17.4s\n",
      "[CV 8/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  17.6s\n",
      "[CV 2/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  17.5s\n",
      "[CV 3/12] END classifier__l1_ratio=0.2, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  17.8s\n",
      "[CV 12/12] END classifier__l1_ratio=0.3, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  18.1s\n",
      "[CV 3/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=   9.5s\n",
      "[CV 11/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  18.2s\n",
      "[CV 4/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  18.3s\n",
      "[CV 1/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  10.0s\n",
      "[CV 10/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  18.5s\n",
      "[CV 9/12] END classifier__l1_ratio=0.5, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  18.5s\n",
      "[CV 2/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  10.4s\n",
      "[CV 5/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  10.5s\n",
      "[CV 1/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=   9.6s\n",
      "[CV 8/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  11.4s\n",
      "[CV 12/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  10.9s\n",
      "[CV 2/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  10.6s\n",
      "[CV 4/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.3s\n",
      "[CV 4/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  11.7s\n",
      "[CV 1/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=   9.6s\n",
      "[CV 7/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.8s\n",
      "[CV 6/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.3s\n",
      "[CV 4/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  11.7s\n",
      "[CV 9/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  13.0s\n",
      "[CV 9/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.3s\n",
      "[CV 10/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.5s\n",
      "[CV 5/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  14.4s\n",
      "[CV 7/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.2s\n",
      "[CV 2/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  10.5s\n",
      "[CV 12/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  13.7s\n",
      "[CV 2/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  13.7s\n",
      "[CV 10/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  12.9s\n",
      "[CV 6/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.9s\n",
      "[CV 3/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  12.7s\n",
      "[CV 11/12] END classifier__l1_ratio=0.6, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  17.4s\n",
      "[CV 8/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  13.8s\n",
      "[CV 4/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  11.4s\n",
      "[CV 3/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  16.5s\n",
      "[CV 3/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  14.7s\n",
      "[CV 1/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.9s\n",
      "[CV 5/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  14.4s\n",
      "[CV 11/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.4s\n",
      "[CV 12/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.1s\n",
      "[CV 12/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  11.0s\n",
      "[CV 8/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.4s\n",
      "[CV 4/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.4s\n",
      "[CV 6/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  12.6s\n",
      "[CV 8/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  14.1s\n",
      "[CV 2/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.0s\n",
      "[CV 6/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  15.7s\n",
      "[CV 11/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.5s\n",
      "[CV 11/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  12.6s\n",
      "[CV 5/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.3s\n",
      "[CV 10/12] END classifier__l1_ratio=0.7, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  17.4s\n",
      "[CV 9/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  13.1s\n",
      "[CV 10/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.9s\n",
      "[CV 12/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.5s\n",
      "[CV 10/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  13.2s\n",
      "[CV 11/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  14.8s\n",
      "[CV 5/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  14.0s\n",
      "[CV 9/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.6s\n",
      "[CV 1/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.500) total time=  14.6s\n",
      "[CV 7/12] END classifier__l1_ratio=0.8, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  17.5s\n",
      "[CV 3/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=0.000) total time=  15.2s\n",
      "[CV 9/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.4s\n",
      "[CV 6/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  17.0s\n",
      "[CV 7/12] END classifier__l1_ratio=0.9, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.8s\n",
      "[CV 8/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.0s\n",
      "[CV 7/12] END classifier__l1_ratio=1, classifier__penalty=elasticnet; accuracy: (test=1.000) total time=  16.5s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-5 {color: black;}#sk-container-id-5 pre{padding: 0;}#sk-container-id-5 div.sk-toggleable {background-color: white;}#sk-container-id-5 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-5 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-5 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-5 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-5 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-5 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-5 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-5 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-5 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-5 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-5 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-5 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-5 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-5 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-5 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-5 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-5 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-5 div.sk-item {position: relative;z-index: 1;}#sk-container-id-5 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-5 div.sk-item::before, #sk-container-id-5 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-5 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-5 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-5 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-5 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-5 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-5 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-5 div.sk-label-container {text-align: center;}#sk-container-id-5 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-5 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-5\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;scaler&#x27;, StandardScaler()),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             n_jobs=60,\n",
       "             param_grid={&#x27;classifier__l1_ratio&#x27;: [0.1, 0.2, 0.3, 0.4, 0.5, 0.6,\n",
       "                                                  0.7, 0.8, 0.9, 1],\n",
       "                         &#x27;classifier__penalty&#x27;: [&#x27;elasticnet&#x27;]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)}, verbose=3)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-17\" type=\"checkbox\" ><label for=\"sk-estimator-id-17\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GridSearchCV</label><div class=\"sk-toggleable__content\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;scaler&#x27;, StandardScaler()),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             n_jobs=60,\n",
       "             param_grid={&#x27;classifier__l1_ratio&#x27;: [0.1, 0.2, 0.3, 0.4, 0.5, 0.6,\n",
       "                                                  0.7, 0.8, 0.9, 1],\n",
       "                         &#x27;classifier__penalty&#x27;: [&#x27;elasticnet&#x27;]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)}, verbose=3)</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-18\" type=\"checkbox\" ><label for=\"sk-estimator-id-18\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;scaler&#x27;, StandardScaler()),\n",
       "                (&#x27;classifier&#x27;,\n",
       "                 LogisticRegression(max_iter=10000, random_state=42,\n",
       "                                    solver=&#x27;saga&#x27;))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-19\" type=\"checkbox\" ><label for=\"sk-estimator-id-19\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScaler</label><div class=\"sk-toggleable__content\"><pre>StandardScaler()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-20\" type=\"checkbox\" ><label for=\"sk-estimator-id-20\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">LogisticRegression</label><div class=\"sk-toggleable__content\"><pre>LogisticRegression(max_iter=10000, random_state=42, solver=&#x27;saga&#x27;)</pre></div></div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[('scaler', StandardScaler()),\n",
       "                                       ('classifier',\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           random_state=42,\n",
       "                                                           solver='saga'))]),\n",
       "             n_jobs=60,\n",
       "             param_grid={'classifier__l1_ratio': [0.1, 0.2, 0.3, 0.4, 0.5, 0.6,\n",
       "                                                  0.7, 0.8, 0.9, 1],\n",
       "                         'classifier__penalty': ['elasticnet']},\n",
       "             refit='accuracy',\n",
       "             scoring={'accuracy': make_scorer(accuracy_score)}, verbose=3)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grid_search.fit(abdelaal_data, abdelaal_labels_simple)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Parameters: {'classifier__l1_ratio': 0.1, 'classifier__penalty': 'elasticnet'}\n",
      "Best CV average accuracy: 0.667\n"
     ]
    }
   ],
   "source": [
    "non_zero_counts = []\n",
    "for params, mean_score in zip(grid_search.cv_results_['params'], grid_search.cv_results_['mean_test_accuracy']):\n",
    "    # Manually fit the pipeline with the parameters\n",
    "    model = log_pipe.set_params(**params)\n",
    "    model.fit(abdelaal_data, abdelaal_labels_simple)  # Re-fit the model on the entire dataset\n",
    "    non_zero_count = np.sum(model.named_steps['classifier'].coef_ != 0)\n",
    "    non_zero_counts.append((params, mean_score, non_zero_count))\n",
    "\n",
    "non_zero_counts_df = pd.DataFrame(non_zero_counts)\n",
    "\n",
    "non_zero_counts_df\n",
    "\n",
    "non_zero_counts_df = non_zero_counts_df.rename(columns={0: \"Paramaters\", 1: \"Average Accuracy\", 2: \"Non-zero coefficeints\"})\n",
    "\n",
    "\n",
    "# Look at best paramaters and accuracy and save results to a data frame\n",
    "print(\"Best Parameters:\", grid_search.best_params_)\n",
    "print(f\"Best CV average accuracy: {grid_search.best_score_:.3f}\")\n",
    "results_genes = pd.concat([pd.DataFrame(grid_search.cv_results_[\"params\"]),\n",
    "           pd.DataFrame(grid_search.cv_results_[\"mean_test_accuracy\"], columns = [\"Average Accuracy\"]),\n",
    "           pd.DataFrame(grid_search.cv_results_[\"std_test_accuracy\"], columns=[\"SD accuracy\"])],axis=1)\n",
    "results_genes.sort_values(by='Average Accuracy', inplace=True)\n",
    "results_genes = pd.concat([results_genes, non_zero_counts_df.loc[:,[\"Non-zero coefficeints\"]]], axis=1)\n",
    "results_genes.sort_values(by='Average Accuracy', inplace=True)\n",
    "results_genes\n",
    "\n",
    "results_genes\n",
    "results_models = pd.DataFrame(grid_search.cv_results_)\n",
    "results_models.sort_values(by='rank_test_accuracy', inplace=True)\n",
    "results_genes\n",
    "results_genes.to_csv(\"/home/workspace/jogrady/ML4TB/results/Logistic_regression/CV_results/Abdelaal_gene_CV_results.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>classifier__l1_ratio</th>\n",
       "      <th>classifier__penalty</th>\n",
       "      <th>Average Accuracy</th>\n",
       "      <th>SD accuracy</th>\n",
       "      <th>Non-zero coefficeints</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.7</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.408248</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.8</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.408248</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.9</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.408248</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.5</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.541667</td>\n",
       "      <td>0.431003</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.6</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.541667</td>\n",
       "      <td>0.431003</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.0</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.541667</td>\n",
       "      <td>0.379601</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.4</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.399653</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.3</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.360844</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.1</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.372678</td>\n",
       "      <td>695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.2</td>\n",
       "      <td>elasticnet</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.372678</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   classifier__l1_ratio classifier__penalty  Average Accuracy  SD accuracy  \\\n",
       "6                   0.7          elasticnet          0.500000     0.408248   \n",
       "7                   0.8          elasticnet          0.500000     0.408248   \n",
       "8                   0.9          elasticnet          0.500000     0.408248   \n",
       "4                   0.5          elasticnet          0.541667     0.431003   \n",
       "5                   0.6          elasticnet          0.541667     0.431003   \n",
       "9                   1.0          elasticnet          0.541667     0.379601   \n",
       "3                   0.4          elasticnet          0.583333     0.399653   \n",
       "2                   0.3          elasticnet          0.625000     0.360844   \n",
       "0                   0.1          elasticnet          0.666667     0.372678   \n",
       "1                   0.2          elasticnet          0.666667     0.372678   \n",
       "\n",
       "   Non-zero coefficeints  \n",
       "6                     85  \n",
       "7                     70  \n",
       "8                     61  \n",
       "4                    123  \n",
       "5                     97  \n",
       "9                     52  \n",
       "3                    167  \n",
       "2                    234  \n",
       "0                    695  \n",
       "1                    358  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {'classifier__l1_ratio': 0.2, 'classifier__penalty': 'elasticnet'}\n",
    "model = log_pipe.set_params(**params)\n",
    "model.fit(abdelaal_data, abdelaal_labels_simple)  # Re-fit the model on the entire dataset\n",
    "# Save model in case we ever need it again e.g. for external data\n",
    "with open('/home/workspace/jogrady/ML4TB/work/models/Logistic_regression/Abdelaal_Logistic_regression_l10.2_search.pkl', 'wb') as f:\n",
    "    pickle.dump(model, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variance captured by 9 PCs:0.82\n"
     ]
    }
   ],
   "source": [
    " \n",
    "# Do not need to scale for PCA - not recommended in VST\n",
    "pca = PCA(random_state=88, n_components=24)\n",
    "pca.fit(abdelaal_data)\n",
    "explained_variance = pca.explained_variance_ratio_\n",
    "# Plotting the elbow curve\n",
    "plt.figure(figsize=(15, 6))\n",
    "plt.plot(range(1, 24+1), explained_variance, marker='o')\n",
    "plt.xlabel('Number of Principal Components based on training set')\n",
    "plt.ylabel('Variance explained')\n",
    "plt.title('Elbow Method for Optimal Number of Components in training set')\n",
    "plt.grid()\n",
    "plt.xticks(range(1, 24 + 1))\n",
    "plt.axvline(x=6, color='r', linestyle='--', label='Optimal Components (based on elbow)')\n",
    "plt.axvline(x=9, color='g', linestyle='--', label='80% variance captured')\n",
    "plt.legend()\n",
    "plt.show()\n",
    "print(f'Variance captured by 9 PCs:{pca.explained_variance_ratio_[:9].sum():.2f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-6 {color: black;}#sk-container-id-6 pre{padding: 0;}#sk-container-id-6 div.sk-toggleable {background-color: white;}#sk-container-id-6 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-6 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-6 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-6 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-6 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-6 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-6 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-6 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-6 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-6 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-6 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-6 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-6 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-6 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-6 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-6 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-6 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-6 div.sk-item {position: relative;z-index: 1;}#sk-container-id-6 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-6 div.sk-item::before, #sk-container-id-6 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-6 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-6 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-6 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-6 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-6 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-6 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-6 div.sk-label-container {text-align: center;}#sk-container-id-6 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-6 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-6\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;pca&#x27;, PCA(random_state=42)),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=&#x27;none&#x27;,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             param_grid={&#x27;pca__n_components&#x27;: [1, 2, 3, 4, 5, 6, 7, 8, 9]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)})</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-21\" type=\"checkbox\" ><label for=\"sk-estimator-id-21\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GridSearchCV</label><div class=\"sk-toggleable__content\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;pca&#x27;, PCA(random_state=42)),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=&#x27;none&#x27;,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             param_grid={&#x27;pca__n_components&#x27;: [1, 2, 3, 4, 5, 6, 7, 8, 9]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)})</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-22\" type=\"checkbox\" ><label for=\"sk-estimator-id-22\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;pca&#x27;, PCA(random_state=42)),\n",
       "                (&#x27;classifier&#x27;,\n",
       "                 LogisticRegression(max_iter=10000, penalty=&#x27;none&#x27;,\n",
       "                                    random_state=42, solver=&#x27;saga&#x27;))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-23\" type=\"checkbox\" ><label for=\"sk-estimator-id-23\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">PCA</label><div class=\"sk-toggleable__content\"><pre>PCA(random_state=42)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-24\" type=\"checkbox\" ><label for=\"sk-estimator-id-24\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">LogisticRegression</label><div class=\"sk-toggleable__content\"><pre>LogisticRegression(max_iter=10000, penalty=&#x27;none&#x27;, random_state=42,\n",
       "                   solver=&#x27;saga&#x27;)</pre></div></div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[('pca', PCA(random_state=42)),\n",
       "                                       ('classifier',\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty='none',\n",
       "                                                           random_state=42,\n",
       "                                                           solver='saga'))]),\n",
       "             param_grid={'pca__n_components': [1, 2, 3, 4, 5, 6, 7, 8, 9]},\n",
       "             refit='accuracy',\n",
       "             scoring={'accuracy': make_scorer(accuracy_score)})"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set up the pipeline\n",
    "PCA_Pipeline = Pipeline(steps=[('pca', PCA(random_state=42)),\n",
    "('classifier', LogisticRegression(max_iter=10000, penalty=\"none\", solver='saga', tol=0.0001, random_state=42))])\n",
    "\n",
    "# Set up a grid for each PC\n",
    "pca_param_grid = {'pca__n_components': list(range(1, 10))}\n",
    "\n",
    "# Apply\n",
    "LR_pca_search_model = GridSearchCV(PCA_Pipeline, pca_param_grid, cv=ps, scoring=scoring, refit=\"accuracy\")\n",
    "\n",
    "# Fit\n",
    "LR_pca_search_model.fit(abdelaal_data, abdelaal_labels_simple.ravel())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Paramater: {'pca__n_components': 9}\n",
      "Best Score: 0.75\n"
     ]
    }
   ],
   "source": [
    "# Look at best paramaters and accuracy and save results to a data frame\n",
    "print(f\"Best Paramater:\", LR_pca_search_model.best_params_)\n",
    "print(f\"Best Score: {LR_pca_search_model.best_score_:.2f}\")\n",
    "results_pca = pd.concat([pd.DataFrame(LR_pca_search_model.cv_results_[\"params\"]),\n",
    "           pd.DataFrame(LR_pca_search_model.cv_results_[\"mean_test_accuracy\"], columns = [\"CV Accuracy\"]),\n",
    "           pd.DataFrame(LR_pca_search_model.cv_results_[\"std_test_accuracy\"], columns=[\"SD accuracy\"])],axis=1)\n",
    "results_pca.sort_values(by='CV Accuracy', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/workspace/jogrady/ML4TB/work/models/Logistic_regression/Abdelaal_Logistic_regression_PCA_search.pkl', 'wb') as f:\n",
    "    pickle.dump(LR_pca_search_model, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   pca__n_components  CV Accuracy  SD accuracy\n",
      "0                  1     0.041667     0.138193\n",
      "3                  4     0.416667     0.343592\n",
      "1                  2     0.458333     0.379601\n",
      "2                  3     0.500000     0.353553\n",
      "6                  7     0.500000     0.408248\n",
      "5                  6     0.541667     0.431003\n",
      "4                  5     0.583333     0.343592\n",
      "7                  8     0.625000     0.414578\n",
      "8                  9     0.750000     0.381881\n"
     ]
    }
   ],
   "source": [
    "print(results_pca)\n",
    "results_pca.to_csv(path_or_buf=\"/home/workspace/jogrady/ML4TB/results/Logistic_regression/CV_results/Abdelaal_PCA_CV_results.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Component_1</th>\n",
       "      <th>Component_2</th>\n",
       "      <th>Component_3</th>\n",
       "      <th>Component_4</th>\n",
       "      <th>Component_5</th>\n",
       "      <th>Component_6</th>\n",
       "      <th>Component_7</th>\n",
       "      <th>Component_8</th>\n",
       "      <th>Component_9</th>\n",
       "      <th>Component_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSBTAG00000039256</td>\n",
       "      <td>STEAP4</td>\n",
       "      <td>ENSBTAG00000054494</td>\n",
       "      <td>ENSBTAG00000031825</td>\n",
       "      <td>PRSS2</td>\n",
       "      <td>PRSS2</td>\n",
       "      <td>ENSBTAG00000052325</td>\n",
       "      <td>ENSBTAG00000008328</td>\n",
       "      <td>ADAM11</td>\n",
       "      <td>ENSBTAG00000054947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSBTAG00000038080</td>\n",
       "      <td>SLC1A1</td>\n",
       "      <td>ENSBTAG00000006383</td>\n",
       "      <td>ENSBTAG00000025398</td>\n",
       "      <td>RBM44</td>\n",
       "      <td>ENSBTAG00000040367</td>\n",
       "      <td>ENSBTAG00000011961</td>\n",
       "      <td>ENSBTAG00000048135</td>\n",
       "      <td>ENSBTAG00000050515</td>\n",
       "      <td>CDH2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSBTAG00000013055</td>\n",
       "      <td>ENSBTAG00000049271</td>\n",
       "      <td>ENSBTAG00000031825</td>\n",
       "      <td>ENSBTAG00000038080</td>\n",
       "      <td>ENSBTAG00000051519</td>\n",
       "      <td>FXYD4</td>\n",
       "      <td>ENSBTAG00000053635</td>\n",
       "      <td>PRSS2</td>\n",
       "      <td>WNT5A</td>\n",
       "      <td>ENSBTAG00000052233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALOX15</td>\n",
       "      <td>ENSBTAG00000055309</td>\n",
       "      <td>ENSBTAG00000051896</td>\n",
       "      <td>ENSBTAG00000013055</td>\n",
       "      <td>ENSBTAG00000039256</td>\n",
       "      <td>TUBB1</td>\n",
       "      <td>TUBB1</td>\n",
       "      <td>TMC7</td>\n",
       "      <td>IQSEC3</td>\n",
       "      <td>MRPS25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSBTAG00000006383</td>\n",
       "      <td>ENSBTAG00000053356</td>\n",
       "      <td>KLRF2</td>\n",
       "      <td>TACR3</td>\n",
       "      <td>SYT7</td>\n",
       "      <td>PDLIM1</td>\n",
       "      <td>MPIG6B</td>\n",
       "      <td>ENSBTAG00000019017</td>\n",
       "      <td>ENSBTAG00000054705</td>\n",
       "      <td>OSBP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5517</th>\n",
       "      <td>ENSBTAG00000052428</td>\n",
       "      <td>RBM44</td>\n",
       "      <td>RADX</td>\n",
       "      <td>ENSBTAG00000054947</td>\n",
       "      <td>bta-mir-11979</td>\n",
       "      <td>BATF2</td>\n",
       "      <td>CACNA1B</td>\n",
       "      <td>ENSBTAG00000050510</td>\n",
       "      <td>ENSBTAG00000052514</td>\n",
       "      <td>PGA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5518</th>\n",
       "      <td>SLC8A1</td>\n",
       "      <td>ENSBTAG00000050585</td>\n",
       "      <td>CBLIF</td>\n",
       "      <td>ENSBTAG00000037421</td>\n",
       "      <td>ENSBTAG00000052233</td>\n",
       "      <td>ENSBTAG00000045580</td>\n",
       "      <td>ENSBTAG00000038080</td>\n",
       "      <td>IFITM1</td>\n",
       "      <td>ENSBTAG00000006383</td>\n",
       "      <td>ENSBTAG00000037539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5519</th>\n",
       "      <td>TGFA</td>\n",
       "      <td>CERS6</td>\n",
       "      <td>ENSBTAG00000037421</td>\n",
       "      <td>ENSBTAG00000040367</td>\n",
       "      <td>BOLA-DQA5</td>\n",
       "      <td>TNFAIP2</td>\n",
       "      <td>BOLA-DQA5</td>\n",
       "      <td>ENSBTAG00000045580</td>\n",
       "      <td>LRP11</td>\n",
       "      <td>ENSBTAG00000055185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520</th>\n",
       "      <td>STEAP4</td>\n",
       "      <td>MRPS25</td>\n",
       "      <td>BOLA-DQA5</td>\n",
       "      <td>BOLA-DQA5</td>\n",
       "      <td>ENSBTAG00000009656</td>\n",
       "      <td>ENSBTAG00000054705</td>\n",
       "      <td>CBLIF</td>\n",
       "      <td>RADX</td>\n",
       "      <td>ENSBTAG00000055211</td>\n",
       "      <td>RADX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5521</th>\n",
       "      <td>SLC30A4</td>\n",
       "      <td>ENSBTAG00000036061</td>\n",
       "      <td>ENSBTAG00000009656</td>\n",
       "      <td>ENSBTAG00000009656</td>\n",
       "      <td>ENSBTAG00000054774</td>\n",
       "      <td>ENSBTAG00000052325</td>\n",
       "      <td>ENSBTAG00000009656</td>\n",
       "      <td>FOLH1B</td>\n",
       "      <td>ARHGEF5</td>\n",
       "      <td>TENM4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5522 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Component_1         Component_2         Component_3  \\\n",
       "0     ENSBTAG00000039256              STEAP4  ENSBTAG00000054494   \n",
       "1     ENSBTAG00000038080              SLC1A1  ENSBTAG00000006383   \n",
       "2     ENSBTAG00000013055  ENSBTAG00000049271  ENSBTAG00000031825   \n",
       "3                 ALOX15  ENSBTAG00000055309  ENSBTAG00000051896   \n",
       "4     ENSBTAG00000006383  ENSBTAG00000053356               KLRF2   \n",
       "...                  ...                 ...                 ...   \n",
       "5517  ENSBTAG00000052428               RBM44                RADX   \n",
       "5518              SLC8A1  ENSBTAG00000050585               CBLIF   \n",
       "5519                TGFA               CERS6  ENSBTAG00000037421   \n",
       "5520              STEAP4              MRPS25           BOLA-DQA5   \n",
       "5521             SLC30A4  ENSBTAG00000036061  ENSBTAG00000009656   \n",
       "\n",
       "             Component_4         Component_5         Component_6  \\\n",
       "0     ENSBTAG00000031825               PRSS2               PRSS2   \n",
       "1     ENSBTAG00000025398               RBM44  ENSBTAG00000040367   \n",
       "2     ENSBTAG00000038080  ENSBTAG00000051519               FXYD4   \n",
       "3     ENSBTAG00000013055  ENSBTAG00000039256               TUBB1   \n",
       "4                  TACR3                SYT7              PDLIM1   \n",
       "...                  ...                 ...                 ...   \n",
       "5517  ENSBTAG00000054947       bta-mir-11979               BATF2   \n",
       "5518  ENSBTAG00000037421  ENSBTAG00000052233  ENSBTAG00000045580   \n",
       "5519  ENSBTAG00000040367           BOLA-DQA5             TNFAIP2   \n",
       "5520           BOLA-DQA5  ENSBTAG00000009656  ENSBTAG00000054705   \n",
       "5521  ENSBTAG00000009656  ENSBTAG00000054774  ENSBTAG00000052325   \n",
       "\n",
       "             Component_7         Component_8         Component_9  \\\n",
       "0     ENSBTAG00000052325  ENSBTAG00000008328              ADAM11   \n",
       "1     ENSBTAG00000011961  ENSBTAG00000048135  ENSBTAG00000050515   \n",
       "2     ENSBTAG00000053635               PRSS2               WNT5A   \n",
       "3                  TUBB1                TMC7              IQSEC3   \n",
       "4                 MPIG6B  ENSBTAG00000019017  ENSBTAG00000054705   \n",
       "...                  ...                 ...                 ...   \n",
       "5517             CACNA1B  ENSBTAG00000050510  ENSBTAG00000052514   \n",
       "5518  ENSBTAG00000038080              IFITM1  ENSBTAG00000006383   \n",
       "5519           BOLA-DQA5  ENSBTAG00000045580               LRP11   \n",
       "5520               CBLIF                RADX  ENSBTAG00000055211   \n",
       "5521  ENSBTAG00000009656              FOLH1B             ARHGEF5   \n",
       "\n",
       "            Component_10  \n",
       "0     ENSBTAG00000054947  \n",
       "1                   CDH2  \n",
       "2     ENSBTAG00000052233  \n",
       "3                 MRPS25  \n",
       "4                  OSBP2  \n",
       "...                  ...  \n",
       "5517                PGA5  \n",
       "5518  ENSBTAG00000037539  \n",
       "5519  ENSBTAG00000055185  \n",
       "5520                RADX  \n",
       "5521               TENM4  \n",
       "\n",
       "[5522 rows x 10 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'residuals_all.columns' contains the names of the genes\n",
    "\n",
    "pca = PCA(random_state=42, n_components=10)\n",
    "pca_fit = pca.fit(abdelaal_data)\n",
    "comp_genes= []\n",
    "pca_genes = pd.DataFrame()\n",
    "pca_results = pca_fit.fit_transform(abdelaal_data)\n",
    "colors = ['steelblue' if label == 0 else 'crimson' for label in abdelaal_labels_simple.ravel()]\n",
    "plt.scatter(pca_results[:, 0], pca_results[:, 1], label=\"Training data\", c=colors, s=20)\n",
    "handles = [plt.Line2D([0], [0], marker='s', color='w', markerfacecolor='steelblue', markersize=5, label='Control'),\n",
    "           plt.Line2D([0], [0], marker='s', color='w', markerfacecolor='crimson', markersize=5, label='Infected')]\n",
    "plt.title(\"PCA Visualization of eigenvectors\", fontweight='bold')\n",
    "plt.legend(handles=handles)\n",
    "plt.xlabel(\"PC1\")\n",
    "plt.ylabel(\"PC2\")\n",
    "plt.show()\n",
    "# Iterate through each component\n",
    "for component_idx in range(10):\n",
    "    # Calculate loadings for the current component\n",
    "    loadings = pca_fit.components_.T * np.sqrt(pca_fit.explained_variance_)\n",
    "\n",
    "    # Sort genes by their influence on the current component\n",
    "    sorted_genes = abdelaal_data.columns[np.argsort(loadings[:, component_idx])]\n",
    "\n",
    "    # Visualize the top N genes that contribute the most to the current component\n",
    "    N = pca_fit.components_.shape[1]\n",
    "    top_genes = sorted_genes[-N:]\n",
    "\n",
    "    pca_genes[component_idx] = top_genes[::-1]\n",
    "    comp_genes = np.append(comp_genes, top_genes)\n",
    "pca_genes.columns = [\"Component_1\",\"Component_2\",\"Component_3\",\"Component_4\",\n",
    "                     \"Component_5\",\"Component_6\",\"Component_7\",\"Component_8\",\n",
    "                     \"Component_9\",\"Component_10\"]                      \n",
    "pca_genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up the grid for subplots (10 rows, 3 columns)\n",
    "fig, axes = plt.subplots(3, 3, figsize=(18, 30))\n",
    "fig.suptitle(\"PCA Visualization of Eigenvectors\", fontweight='bold', fontsize=16)\n",
    "\n",
    "# Flatten the axes array for easy iteration\n",
    "axes = axes.flatten()\n",
    "\n",
    "# Loop to generate scatter plots from PC1 vs each PC up to PC30\n",
    "for i in range(1, 10):\n",
    "    ax = axes[i-1]\n",
    "    ax.scatter(pca_results[:, 0], pca_results[:, i], c=colors, s=20)\n",
    "    ax.set_title(f\"PC1 vs PC{i+1}\")\n",
    "    ax.set_xlabel(\"PC1\")\n",
    "    ax.set_ylabel(f\"PC{i+1}\")\n",
    "\n",
    "# Custom legend only in the first subplot (for cleanliness)\n",
    "handles = [\n",
    "    plt.Line2D([0], [0], marker='s', color='w', markerfacecolor='steelblue', markersize=5, label='Control'),\n",
    "    plt.Line2D([0], [0], marker='s', color='w', markerfacecolor='crimson', markersize=5, label='Infected')\n",
    "]\n",
    "axes[0].legend(handles=handles)\n",
    "\n",
    "# Hide any unused subplots (since there are only 29 plots for a 30-slot grid)\n",
    "for j in range(29, len(axes)):\n",
    "    axes[j].axis('off')\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.96])  # Adjust layout to fit title\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>coef</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Component_9</td>\n",
       "      <td>[-1.536679067820141]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Component_3</td>\n",
       "      <td>[-0.9522225560500457]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Component_7</td>\n",
       "      <td>[-0.5054857892466998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Component_6</td>\n",
       "      <td>[-0.3839239904971617]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Component_1</td>\n",
       "      <td>[-0.10098649834738362]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Component_4</td>\n",
       "      <td>[0.1532031343030736]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Component_2</td>\n",
       "      <td>[0.2458453043305297]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Component_8</td>\n",
       "      <td>[0.29649556847883085]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Component_5</td>\n",
       "      <td>[0.44264740413875125]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      features                    coef\n",
       "8  Component_9    [-1.536679067820141]\n",
       "2  Component_3   [-0.9522225560500457]\n",
       "6  Component_7   [-0.5054857892466998]\n",
       "5  Component_6   [-0.3839239904971617]\n",
       "0  Component_1  [-0.10098649834738362]\n",
       "3  Component_4    [0.1532031343030736]\n",
       "1  Component_2    [0.2458453043305297]\n",
       "7  Component_8   [0.29649556847883085]\n",
       "4  Component_5   [0.44264740413875125]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Feature_weight_pca = pd.DataFrame(zip(pca_genes.columns, np.transpose(LR_pca_search_model.best_estimator_.named_steps[\"classifier\"].coef_)), columns=['features', 'coef'])#.sort_values(by='coef', inplace=True)\n",
    "Feature_weight_pca.sort_values(by='coef', inplace = True)\n",
    "Feature_weight_pca = Feature_weight_pca.loc[(Feature_weight_pca != 0).all(axis=1), :]\n",
    "Feature_weight_pca\n",
    "# PC 5 is very important in the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Component  T-Statistic   P-Value\n",
      "0         0    -0.033151  0.973867\n",
      "1         1     1.572100  0.130873\n",
      "2         2    -1.988008  0.060004\n",
      "3         3     0.243318  0.810120\n",
      "4         4    -0.297891  0.768713\n",
      "5         5    -2.306678  0.031360\n",
      "6         6     0.744525  0.464809\n",
      "7         7     0.191887  0.849674\n",
      "8         8    -0.175597  0.862293\n",
      "9         9    -0.562930  0.579444\n",
      "Components with significantly different means:\n",
      "Empty DataFrame\n",
      "Columns: [Component, T-Statistic, P-Value, Corrected P-Value, Significant (BH)]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "df_pca = pd.DataFrame(pca_results)\n",
    "Control = df_pca.iloc[:8, :] # 45 samples in control\n",
    "Infected = df_pca.iloc[9:, :] # 42 in infected\n",
    "\n",
    "# Perform t-test for each component\n",
    "results = pd.DataFrame(columns=['Component', 'T-Statistic', 'P-Value'])\n",
    "\n",
    "for column in df_pca.columns:\n",
    "    t_statistic, p_value = stats.ttest_ind(Control[column], Infected[column])\n",
    "    \n",
    "    new_row = pd.DataFrame({'Component': [column], 'T-Statistic': [t_statistic], 'P-Value': [p_value]})\n",
    "    results = pd.concat([results, new_row], ignore_index=True)\n",
    "\n",
    "# Print the results\n",
    "print(results)\n",
    "\n",
    "# Identify components with significantly different means\n",
    "\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "reject, pvals_corrected, _, _ = multipletests(results[\"P-Value\"], alpha=0.05, method='fdr_bh')\n",
    "results['Corrected P-Value'] = pvals_corrected\n",
    "results['Significant (BH)'] = reject \n",
    "\n",
    "\n",
    "significantly_different = results[results['Corrected P-Value'] < 0.05]\n",
    "print(\"Components with significantly different means:\")\n",
    "print(significantly_different)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. ICA\n",
    "\n",
    "ICA and PCA are similar to each other however ICA attemps to transform the data into statistically significant non-Gaussian components. Often times this is estimated by Kurtosis but the number of  !'PCs'! used for the whitening procedure is often those that reach 80,90,95% etc, hence we will use this for below  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ICA\n",
    "from sklearn.decomposition import FastICA\n",
    "ICA_transformer = FastICA(n_components=9, # from PCA - 80% of variance\n",
    "        random_state=42,\n",
    "        max_iter=1000, tol=0.0001,\n",
    "        whiten='unit-variance')\n",
    "\n",
    "df_train_ica = ICA_transformer.fit_transform(abdelaal_data)\n",
    "kurtosis_scores = [kurtosis(df_train_ica[:, i]) for i in range(df_train_ica.shape[1])]\n",
    "n_components = np.argmax(kurtosis_scores) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of non-normal components in training data = 3 - however we will use the same as above for PCA\n",
    "n_components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-7 {color: black;}#sk-container-id-7 pre{padding: 0;}#sk-container-id-7 div.sk-toggleable {background-color: white;}#sk-container-id-7 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-7 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-7 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-7 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-7 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-7 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-7 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-7 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-7 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-7 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-7 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-7 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-7 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-7 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-7 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-7 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-7 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-7 div.sk-item {position: relative;z-index: 1;}#sk-container-id-7 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-7 div.sk-item::before, #sk-container-id-7 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-7 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-7 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-7 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-7 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-7 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-7 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-7 div.sk-label-container {text-align: center;}#sk-container-id-7 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-7 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-7\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;ica&#x27;,\n",
       "                                        FastICA(max_iter=5000,\n",
       "                                                random_state=42)),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=&#x27;none&#x27;,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             param_grid={&#x27;ica__n_components&#x27;: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)})</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-25\" type=\"checkbox\" ><label for=\"sk-estimator-id-25\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GridSearchCV</label><div class=\"sk-toggleable__content\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;ica&#x27;,\n",
       "                                        FastICA(max_iter=5000,\n",
       "                                                random_state=42)),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=&#x27;none&#x27;,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             param_grid={&#x27;ica__n_components&#x27;: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)})</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-26\" type=\"checkbox\" ><label for=\"sk-estimator-id-26\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;ica&#x27;, FastICA(max_iter=5000, random_state=42)),\n",
       "                (&#x27;classifier&#x27;,\n",
       "                 LogisticRegression(max_iter=10000, penalty=&#x27;none&#x27;,\n",
       "                                    random_state=42, solver=&#x27;saga&#x27;))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-27\" type=\"checkbox\" ><label for=\"sk-estimator-id-27\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">FastICA</label><div class=\"sk-toggleable__content\"><pre>FastICA(max_iter=5000, random_state=42)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-28\" type=\"checkbox\" ><label for=\"sk-estimator-id-28\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">LogisticRegression</label><div class=\"sk-toggleable__content\"><pre>LogisticRegression(max_iter=10000, penalty=&#x27;none&#x27;, random_state=42,\n",
       "                   solver=&#x27;saga&#x27;)</pre></div></div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[('ica',\n",
       "                                        FastICA(max_iter=5000,\n",
       "                                                random_state=42)),\n",
       "                                       ('classifier',\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty='none',\n",
       "                                                           random_state=42,\n",
       "                                                           solver='saga'))]),\n",
       "             param_grid={'ica__n_components': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},\n",
       "             refit='accuracy',\n",
       "             scoring={'accuracy': make_scorer(accuracy_score)})"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set up the pipeline\n",
    "ICA_Pipeline = Pipeline(steps=[('ica', FastICA( \n",
    "        random_state=42,\n",
    "        max_iter=5000, tol=0.0001,\n",
    "        whiten='unit-variance')),\n",
    "('classifier', LogisticRegression(max_iter=10000, penalty=\"none\", solver='saga', tol=0.0001, random_state=42))])\n",
    "\n",
    "# Set up grid of components - 80% from PCA above\n",
    "ica_param_grid = {'ica__n_components': list(range(1, 11))}\n",
    "\n",
    "\n",
    "# Apply\n",
    "LR_ica_search_model = GridSearchCV(ICA_Pipeline, ica_param_grid, cv=ps, scoring=scoring, refit=\"accuracy\")\n",
    "\n",
    "\n",
    "# Fit\n",
    "LR_ica_search_model.fit(abdelaal_data, abdelaal_labels_simple.ravel())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Parameters: {'ica__n_components': 9}\n",
      "Best Score: 0.71\n",
      "   ica__n_components  CV Accuracy  SD accuracy\n",
      "0                  1     0.000000     0.000000\n",
      "1                  2     0.208333     0.246503\n",
      "6                  7     0.458333     0.431003\n",
      "5                  6     0.500000     0.456435\n",
      "2                  3     0.541667     0.379601\n",
      "3                  4     0.541667     0.379601\n",
      "4                  5     0.583333     0.343592\n",
      "9                 10     0.583333     0.399653\n",
      "7                  8     0.625000     0.414578\n",
      "8                  9     0.708333     0.379601\n"
     ]
    }
   ],
   "source": [
    "# Look at best paramaters and accuracy and save results to a data frame\n",
    "print(\"Best Parameters:\", LR_ica_search_model.best_params_)\n",
    "print(f\"Best Score: {LR_ica_search_model.best_score_:.2f}\")\n",
    "results_ica = pd.concat([pd.DataFrame(LR_ica_search_model.cv_results_[\"params\"]),\n",
    "           pd.DataFrame(LR_ica_search_model.cv_results_[\"mean_test_accuracy\"], columns = [\"CV Accuracy\"]),\n",
    "           pd.DataFrame(LR_ica_search_model.cv_results_[\"std_test_accuracy\"], columns=[\"SD accuracy\"])],axis=1)\n",
    "results_ica.sort_values(by='CV Accuracy', inplace=True)\n",
    "\n",
    "with open('/home/workspace/jogrady/ML4TB/work/models/Logistic_regression/Abdelaal_Logistic_regression_ICA_search.pkl', 'wb') as f:\n",
    "    pickle.dump(LR_ica_search_model, f)\n",
    "    \n",
    "print(results_ica)\n",
    "results_ica.to_csv(path_or_buf=\"/home/workspace/jogrady/ML4TB/results/Logistic_regression/CV_results/Abdelaal_ICA_CV_results.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Non-negative matrix factorization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-8 {color: black;}#sk-container-id-8 pre{padding: 0;}#sk-container-id-8 div.sk-toggleable {background-color: white;}#sk-container-id-8 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-8 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-8 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-8 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-8 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-8 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-8 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-8 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-8 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-8 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-8 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-8 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-8 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-8 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-8 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-8 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-8 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-8 div.sk-item {position: relative;z-index: 1;}#sk-container-id-8 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-8 div.sk-item::before, #sk-container-id-8 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-8 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-8 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-8 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-8 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-8 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-8 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-8 div.sk-label-container {text-align: center;}#sk-container-id-8 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-8 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-8\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;nmf&#x27;,\n",
       "                                        NMF(init=&#x27;random&#x27;, max_iter=10000,\n",
       "                                            random_state=42)),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=None,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             n_jobs=10,\n",
       "             param_grid={&#x27;nmf__n_components&#x27;: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)})</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-29\" type=\"checkbox\" ><label for=\"sk-estimator-id-29\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GridSearchCV</label><div class=\"sk-toggleable__content\"><pre>GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[(&#x27;nmf&#x27;,\n",
       "                                        NMF(init=&#x27;random&#x27;, max_iter=10000,\n",
       "                                            random_state=42)),\n",
       "                                       (&#x27;classifier&#x27;,\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=None,\n",
       "                                                           random_state=42,\n",
       "                                                           solver=&#x27;saga&#x27;))]),\n",
       "             n_jobs=10,\n",
       "             param_grid={&#x27;nmf__n_components&#x27;: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},\n",
       "             refit=&#x27;accuracy&#x27;,\n",
       "             scoring={&#x27;accuracy&#x27;: make_scorer(accuracy_score)})</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-30\" type=\"checkbox\" ><label for=\"sk-estimator-id-30\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;nmf&#x27;, NMF(init=&#x27;random&#x27;, max_iter=10000, random_state=42)),\n",
       "                (&#x27;classifier&#x27;,\n",
       "                 LogisticRegression(max_iter=10000, penalty=None,\n",
       "                                    random_state=42, solver=&#x27;saga&#x27;))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-31\" type=\"checkbox\" ><label for=\"sk-estimator-id-31\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">NMF</label><div class=\"sk-toggleable__content\"><pre>NMF(init=&#x27;random&#x27;, max_iter=10000, random_state=42)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-32\" type=\"checkbox\" ><label for=\"sk-estimator-id-32\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">LogisticRegression</label><div class=\"sk-toggleable__content\"><pre>LogisticRegression(max_iter=10000, penalty=None, random_state=42, solver=&#x27;saga&#x27;)</pre></div></div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "GridSearchCV(cv=PredefinedSplit(test_fold=array([ 0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7,  8,\n",
       "        8,  9,  9, 10, 10, 11, 11])),\n",
       "             estimator=Pipeline(steps=[('nmf',\n",
       "                                        NMF(init='random', max_iter=10000,\n",
       "                                            random_state=42)),\n",
       "                                       ('classifier',\n",
       "                                        LogisticRegression(max_iter=10000,\n",
       "                                                           penalty=None,\n",
       "                                                           random_state=42,\n",
       "                                                           solver='saga'))]),\n",
       "             n_jobs=10,\n",
       "             param_grid={'nmf__n_components': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},\n",
       "             refit='accuracy',\n",
       "             scoring={'accuracy': make_scorer(accuracy_score)})"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NMF_Pipeline = Pipeline(steps=[('nmf', NMF(\n",
    "    init=\"random\", solver='cd', \n",
    "    beta_loss='frobenius', tol=0.0001, \n",
    "    max_iter=10000, random_state=42, verbose=0)),\n",
    "('classifier', LogisticRegression(max_iter=10000, penalty=None, solver='saga', tol=0.0001, random_state=42))])\n",
    "\n",
    "nmf_param_grid = {'nmf__n_components': list(range(1, 11))}\n",
    "\n",
    "LR_nmf_search_model = GridSearchCV(NMF_Pipeline, nmf_param_grid, cv=ps, n_jobs= 10, scoring=scoring, refit=\"accuracy\")\n",
    "\n",
    "LR_nmf_search_model.fit(abdelaal_data, abdelaal_labels_simple.ravel())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best Parameters: {'nmf__n_components': 9}\n",
      "Best Score: 0.62\n",
      "   nmf__n_components  CV Accuracy  SD accuracy\n",
      "0                  1     0.000000     0.000000\n",
      "1                  2     0.000000     0.000000\n",
      "2                  3     0.291667     0.320048\n",
      "7                  8     0.416667     0.448764\n",
      "4                  5     0.500000     0.353553\n",
      "6                  7     0.500000     0.456435\n",
      "9                 10     0.500000     0.353553\n",
      "3                  4     0.541667     0.379601\n",
      "5                  6     0.583333     0.343592\n",
      "8                  9     0.625000     0.414578\n"
     ]
    }
   ],
   "source": [
    "# Look at best paramaters and accuracy and save results to a data frame\n",
    "print(\"Best Parameters:\", LR_nmf_search_model.best_params_)\n",
    "print(f\"Best Score: {LR_nmf_search_model.best_score_:.2f}\")\n",
    "results_nmf = pd.concat([pd.DataFrame(LR_nmf_search_model.cv_results_[\"params\"]),\n",
    "           pd.DataFrame(LR_nmf_search_model.cv_results_[\"mean_test_accuracy\"], columns = [\"CV Accuracy\"]),\n",
    "           pd.DataFrame(LR_nmf_search_model.cv_results_[\"std_test_accuracy\"], columns=[\"SD accuracy\"])],axis=1)\n",
    "results_nmf.sort_values(by='CV Accuracy', inplace=True)\n",
    "\n",
    "print(results_nmf)\n",
    "results_nmf.to_csv(path_or_buf=\"/home/workspace/jogrady/ML4TB/results/Logistic_regression/CV_results/Abdelaal_NMF_CV_results.txt\", sep = \"\\t\")\n",
    "with open('/home/workspace/jogrady/ML4TB/work/models/Logistic_regression/Abdelaal_Logistic_regression_NMF_search.pkl', 'wb') as f:\n",
    "    pickle.dump(LR_nmf_search_model, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluation\n",
    "# Ogrady_data_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensemble = pd.read_csv(\n",
    "    \"/home/workspace/jogrady/eqtl_study/eqtl_nextflow/data/RNA_seq/Bos_taurus.ARS-UCD1.2.110.gtf\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    comment=\"#\",\n",
    ")\n",
    "# Filter for rows where column 3 (V3) is \"gene\"\n",
    "ensemble = ensemble[ensemble[2] == \"gene\"]\n",
    "\n",
    "# Split column 9 (V9) into separate columns for gene_id, gene_version, and gene_name\n",
    "ensemble_split = ensemble[8].str.split(\";\", expand=True)\n",
    "ensemble[\"gene_id\"] = ensemble_split[0].str.replace(\"^gene_id \", \"\", regex=True).str.replace('\"', '', regex=False)\n",
    "ensemble[\"gene_version\"] = ensemble_split[1]\n",
    "ensemble[\"gene_name\"] = ensemble_split[2].str.replace(\"gene_name \", \"\", regex=True).str.replace('\"', '', regex=False)\n",
    "\n",
    "# Clean up the gene_name column\n",
    "ensemble[\"gene_name\"] = (\n",
    "    ensemble[\"gene_name\"]\n",
    "    .str.replace(\"gene_source \", \"\", regex=False)\n",
    "    .str.strip()\n",
    ")\n",
    "ensemble[\"gene_name\"] = np.where(\n",
    "    ensemble[\"gene_name\"].isin([\"ensembl\", \"5S_rRNA\"]), \n",
    "    ensemble[\"gene_id\"], \n",
    "    ensemble[\"gene_name\"]\n",
    ")\n",
    "\n",
    "# Rename columns\n",
    "ensemble.rename(columns={0: \"chr\"}, inplace=True)\n",
    "ensemble = ensemble[[\"gene_id\", \"gene_name\", \"chr\", 3]]\n",
    "ensemble.rename(columns={3: \"pos\"}, inplace=True)\n",
    "\n",
    "# Select only gene_id and gene_name\n",
    "ensemble = ensemble[[\"gene_id\", \"gene_name\"]]\n",
    "\n",
    "# Handle duplicated gene_name values\n",
    "ensemble[\"gene_name\"] = np.where(\n",
    "    ensemble[\"gene_name\"].duplicated(), \n",
    "    ensemble[\"gene_id\"], \n",
    "    ensemble[\"gene_name\"]\n",
    ")\n",
    "\n",
    "# Clean up gene_id and gene_name columns for alignment\n",
    "ensemble[\"gene_name\"] = ensemble[\"gene_name\"].str.replace(' ', '', regex=False)\n",
    "ensemble[\"gene_id\"] = ensemble[\"gene_id\"].str.replace(' ', '', regex=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "ogrady_test = pd.read_csv(\"/home/workspace/jogrady/ML4TB/work/normalisation/vst_combined/ogradytest_abdelaal_vst_normalised_data.txt\", sep = \"\\t\")\n",
    "\n",
    "tested_genes = pd.DataFrame(ogrady_test.index)\n",
    "tested_genes.columns = [\"gene_id\"]\n",
    "\n",
    "tested_genes = tested_genes.merge(ensemble, how=\"left\", on=\"gene_id\")\n",
    "ogrady_test.index = tested_genes[\"gene_name\"].to_numpy()\n",
    "ogrady_test = ogrady_test.T\n",
    "\n",
    "\n",
    "test_labels = pd.read_csv(\"/home/workspace/jogrady/ML4TB/work/normalisation/vst_individual/Test_labels.txt\", sep = \"\\t\").T.to_numpy() \n",
    "test_labels = np.where(test_labels == \"Control\", 0, np.where(test_labels == \"Infected\", 1, test_labels))\n",
    "test_labels = test_labels.astype(int)\n",
    "# Calcualte variances for VST normalised genes\n",
    "variances = abdelaal_data_raw.var(axis=0)\n",
    "# take top 20% and filter\n",
    "threshold = variances.quantile(.80) \n",
    "genes = variances > threshold\n",
    "genes= genes.loc[genes==True].index\n",
    "#print(genes)\n",
    "ogrady_test = ogrady_test.filter(items = genes, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_names = ['Control', 'Infected']\n",
    "report = classification_report(test_labels.ravel(), model.predict(ogrady_test), target_names=target_names, output_dict=True)\n",
    "df_report = pd.DataFrame(report).transpose()\n",
    "Gene_cm_display = ConfusionMatrixDisplay(confusion_matrix = confusion_matrix(test_labels.ravel(),model.predict(ogrady_test)), display_labels = [\"Control\", \"Infected\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "     Control       0.00      0.00      0.00        18\n",
      "    Infected       0.50      1.00      0.67        18\n",
      "\n",
      "    accuracy                           0.50        36\n",
      "   macro avg       0.25      0.50      0.33        36\n",
      "weighted avg       0.25      0.50      0.33        36\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(classification_report(test_labels.ravel(), model.predict(ogrady_test), target_names=target_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "     Control       0.48      0.89      0.63        18\n",
      "    Infected       0.33      0.06      0.10        18\n",
      "\n",
      "    accuracy                           0.47        36\n",
      "   macro avg       0.41      0.47      0.36        36\n",
      "weighted avg       0.41      0.47      0.36        36\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Performance on Test data\n",
    "print(classification_report(test_labels.ravel(), LR_pca_search_model.predict(ogrady_test), target_names=target_names))\n",
    "PCA_cm_display = ConfusionMatrixDisplay(confusion_matrix = confusion_matrix(test_labels.ravel(),LR_pca_search_model.predict(ogrady_test)), display_labels = [\"Control\", \"Infected\"])\n",
    "#ogrady_master_df_predictions = pd.concat([ogrady_master_df_predictions, pd.DataFrame({\"Logistic_PCA_prediction\": LR_pca_search_model.predict(ogrady_data)})], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "     Control       0.48      0.89      0.63        18\n",
      "    Infected       0.33      0.06      0.10        18\n",
      "\n",
      "    accuracy                           0.47        36\n",
      "   macro avg       0.41      0.47      0.36        36\n",
      "weighted avg       0.41      0.47      0.36        36\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Performance on Test data\n",
    "print(classification_report(test_labels.ravel(), LR_ica_search_model.predict(ogrady_test), target_names=target_names))\n",
    "PCA_cm_display = ConfusionMatrixDisplay(confusion_matrix = confusion_matrix(test_labels.ravel(),LR_ica_search_model.predict(ogrady_test)), display_labels = [\"Control\", \"Infected\"])\n",
    "#ogrady_master_df_predictions = pd.concat([ogrady_master_df_predictions, pd.DataFrame({\"Logistic_PCA_prediction\": LR_pca_search_model.predict(test_data)})], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "     Control       0.49      0.94      0.64        18\n",
      "    Infected       0.00      0.00      0.00        18\n",
      "\n",
      "    accuracy                           0.47        36\n",
      "   macro avg       0.24      0.47      0.32        36\n",
      "weighted avg       0.24      0.47      0.32        36\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Performance on Test data\n",
    "print(classification_report(test_labels.ravel(), LR_nmf_search_model.predict(ogrady_test), target_names=target_names))\n",
    "PCA_cm_display = ConfusionMatrixDisplay(confusion_matrix = confusion_matrix(test_labels.ravel(),LR_nmf_search_model.predict(ogrady_test)), display_labels = [\"Control\", \"Infected\"])\n",
    "#ogrady_master_df_predictions = pd.concat([ogrady_master_df_predictions, pd.DataFrame({\"Logistic_PCA_prediction\": LR_pca_search_model.predict(test_data)})], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# predict continuous value\n",
    "# For different decision thresholds\n",
    "y_score_gene = model.predict_proba(ogrady_test)\n",
    "fprG_gene, tprG_gene, t_gene = roc_curve(test_labels.ravel(), y_score_gene[:,1])\n",
    "roc_aucG_gene = auc(fprG_gene, tprG_gene)\n",
    "\n",
    "\n",
    "y_score_pca = LR_pca_search_model.predict_proba(ogrady_test)\n",
    "fprG_pca, tprG_pca, t_pca = roc_curve(test_labels.ravel(), y_score_pca[:,1])\n",
    "roc_aucG_pca = auc(fprG_pca, tprG_pca)\n",
    "\n",
    "\n",
    "y_score_ica = LR_ica_search_model.predict_proba(ogrady_test)\n",
    "fprG_ica, tprG_ica, t_ica = roc_curve(test_labels.ravel(), y_score_ica[:,1])\n",
    "roc_aucG_ica = auc(fprG_ica, tprG_ica)\n",
    "\n",
    "\n",
    "\n",
    "y_score_nmf = LR_nmf_search_model.predict_proba(ogrady_test)\n",
    "fprG_nmf, tprG_nmf, t_nmf = roc_curve(test_labels.ravel(), y_score_nmf[:,1])\n",
    "roc_aucG_nmf = auc(fprG_nmf, tprG_nmf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "matplotlib.use('cairo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%matplotlib inline\n",
    "plt.figure()\n",
    "lw = 2\n",
    "plt.plot(fprG_gene, tprG_gene, color='red',\n",
    "         lw=lw, label='ROC Gene (area = %0.3f)' % roc_aucG_gene)\n",
    "plt.plot(fprG_pca, tprG_pca, color='yellow',\n",
    "         lw=lw, label='ROC PCA (area = %0.3f)' % roc_aucG_pca)\n",
    "plt.plot(fprG_ica, tprG_ica, color='orange',\n",
    "         lw=lw, label='ROC ICA (area = %0.3f)' % roc_aucG_ica)\n",
    "plt.plot(fprG_nmf, tprG_nmf, color='blue',\n",
    "         lw=lw, label='ROC NMF (area = %0.3f)' % roc_aucG_nmf)\n",
    "\n",
    "plt.plot([0, 1], [0, 1], color='black', lw=lw, linestyle='--')\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.ylim([0.0, 1.0])\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "plt.title(\"ROC analysis with different models from abdelaal et al\\non O'Grady et al. hold out set\")\n",
    "plt.legend(loc=\"lower right\")\n",
    "plt.savefig(\"/home/workspace/jogrady/ML4TB/results/Logistic_regression/CV_results/Abdelaal_train_V_OGrady_test_ROC.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
